Change index to start from 1 by default in `predictions.json`

Y-T-G opened the pull request:
Closes https://github.com/ultralytics/ultralytics/issues/18136

The `category_id` in COCO JSON format starts from 1 in general, not just for LVIS. This is also the assumption used in the `convert_coco` function, and hence we subtract 1:
https://github.com/ultralytics/ultralytics/blob/d170f1ab6ad740369e5508788ab267a6639fc851/ultralytics/data/converter.py#L302

<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
--->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
Improved class mapping logic for YOLO evaluation to address indexing inconsistencies in COCO and LVIS datasets.

### 📊 Key Changes  
- Adjusted `class_map` indexing for non-COCO datasets to start from 1 instead of 0.  
- Removed unnecessary indexing condition for LVIS-specific dataset parsing.  
- Fixed category ID offset during bounding box evaluation in OBB (oriented bounding box) results for consistency.

### 🎯 Purpose & Impact  
- 🛠️ **Better compatibility:** Standardizes class indexing across datasets (COCO, LVIS) to ensure evaluation consistency.  
- 🐛 **Bug fixes:** Resolves indexing offset issues, particularly for models handling smaller datasets or custom datasets.  
- 📈 **Improved accuracy:** Ensures proper labeling and metric computation during evaluation, leading to more reliable results for developers and researchers.

Y-T-G committed:
Change index to start from 1 by default

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>

UltralyticsAssistant commented:
👋 Hello @Y-T-G, thank you for submitting a fantastic `ultralytics/ultralytics` 🚀 PR! To help us review and merge your work more efficiently, please ensure the following checklist is complete:

- ✅ **Define a Purpose**: Your detailed PR description is much appreciated! Thank you for linking [the relevant issue](https://github.com/ultralytics/ultralytics/issues/18136) and explaining the context. Excellent work providing clarity on the reasoning behind this change!  
- ✅ **Synchronize with Source**: Please confirm your PR is synchronized with the latest `main` branch of `ultralytics/ultralytics`. If it isn't already up to date, you can sync it by clicking the 'Update branch' button in your PR or by running `git pull` and `git merge main` locally.  
- ✅ **Ensure CI Checks Pass**: Double-check that all [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing for your PR. If any are failing, kindly address the issues.  
- ✅ **Update Documentation**: If necessary, please ensure that the [documentation](https://docs.ultralytics.com) has been updated to reflect your changes for affected datasets like COCO or LVIS.  
- ✅ **Add Tests**: If applicable, include or enhance the tests to verify the behavior of your changes, and ensure all tests are passing locally.  
- ✅ **Sign the CLA**: If this is your first time contributing to Ultralytics, kindly sign our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) by commenting "I have read the CLA Document and I sign the CLA" below.  

🎯 **Precision and Impact:** Your changes streamline the handling of `class_map` for non-COCO datasets and improve alignment with dataset conventions. This is a great contribution for improving consistency in YOLO workflows! Please verify your modifications are as minimalistic as possible to resolve the issue without introducing unrelated changes, as per our [Contributing Guide](https://docs.ultralytics.com/help/contributing). 

If you encounter or identify any edge cases, feel free to share a **minimum reproducible example (MRE)** to help with validation and testing. This helps ensure that all aspects of the fix are covered robustly.  

An Ultralytics engineer will review your PR shortly to assist further. Thank you again for your contribution to Ultralytics! We can’t wait to see this improvement in action. 🚀✨

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher commented:
@Y-T-G thanks for the PR! Could you test val with `main` and this PR on other datasets like VOC that have zero-indexed classes? I'm not sure if this might break something there as I think all of the current YAMLs start from zero, .e.g:

https://github.com/ultralytics/ultralytics/blob/d170f1ab6ad740369e5508788ab267a6639fc851/ultralytics/cfg/datasets/VOC.yaml#L22-L44

Y-T-G committed:
Handle offset in OBB category_id

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>

Y-T-G commented:
@glenn-jocher Sure. I think this only applies to COCO and LVIS because those are the only ones that trigger `eval_json` where we use the COCOAPI and the generated `predictions.json`. And also for DOTA to generate the predictions for submission to DOTA evaluation website. Otherwise, we do not make use of the `predictions.json` ourselves. It just get saved for the users when they use `save_json=True`.

timecollision commented:
I change the code to 
cls = coco80[ann["category_id"]] if cls91to80 else ann["category_id"]  # class
but the "category_id" in predictions.json is still 0

Y-T-G commented:
@timecollision That's for conversion from CCO to YOLO. If you want to test this PR, you can use:

```bash
git clone https://github.com/ultralytics/ultralytics
cd ultralytics
git checkout coco_pred_json
pip install .
```

Then run `model.val(save_json=True)`

Y-T-G commented:
@glenn-jocher I trained a model on VOC for 1 epoch and ran validation with both `main` branch and this PR. The validation results are the same:

## VOC - Detect
**main**
```
results = model.val(save_json=True)
Ultralytics 8.3.48 🚀 Python-3.10.14 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15934MiB)
val: Scanning /datasets/VOC/labels/test2007.cache... 4952 images, 0 backgrounds, 0 corrupt: 100%|██████████| 4952/4952 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 310/310 [00:31<00:00,  9.81it/s]
                   all       4952      12032      0.722      0.662      0.723      0.506
             aeroplane        204        285      0.883       0.67      0.772      0.539
               bicycle        239        337       0.77      0.757      0.817      0.528
                  bird        282        459      0.723      0.699      0.719      0.452
                  boat        172        263      0.628      0.529      0.587      0.369
                bottle        212        469      0.661      0.619      0.671      0.447
                   bus        174        213      0.795      0.744      0.783      0.667
                   car        721       1201      0.779      0.837      0.858      0.642
                   cat        322        358      0.703      0.807      0.793      0.596
                 chair        417        756      0.428      0.541      0.474      0.284
                   cow        127        244      0.792      0.504      0.689      0.482
           diningtable        190        206      0.663      0.733      0.751      0.555
                   dog        418        489      0.848      0.476      0.705      0.517
                 horse        274        348      0.784      0.781      0.832      0.582
             motorbike        222        325      0.742      0.785      0.813      0.546
                person       2007       4528      0.819      0.774      0.844       0.56
           pottedplant        224        480      0.486      0.479      0.466      0.227
                 sheep         97        242      0.668      0.736      0.723      0.549
                  sofa        223        239      0.557      0.665      0.635      0.499
                 train        259        282      0.917      0.548       0.79      0.555
             tvmonitor        229        308      0.804      0.562      0.738      0.526
Speed: 0.1ms preprocess, 2.2ms inference, 0.0ms loss, 0.6ms postprocess per image
```

**PR**
```
results = model.val(save_json=True)
Ultralytics 8.3.48 🚀 Python-3.10.14 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15934MiB)
YOLO11n summary (fused): 238 layers, 2,586,052 parameters, 0 gradients, 6.3 GFLOPs
val: Scanning /datasets/VOC/labels/test2007.cache... 4952 images, 0 backgrounds, 0 corrupt: 100%|██████████| 4952/4952 [00:00<?, ?i
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 310/310 [00:31<00:00,  9
                   all       4952      12032      0.722      0.662      0.723      0.506
             aeroplane        204        285      0.883       0.67      0.772      0.539
               bicycle        239        337       0.77      0.757      0.817      0.528
                  bird        282        459      0.723      0.699      0.719      0.452
                  boat        172        263      0.628      0.529      0.587      0.369
                bottle        212        469      0.661      0.619      0.671      0.447
                   bus        174        213      0.795      0.744      0.783      0.667
                   car        721       1201      0.779      0.837      0.858      0.642
                   cat        322        358      0.703      0.807      0.793      0.596
                 chair        417        756      0.428      0.541      0.474      0.284
                   cow        127        244      0.792      0.504      0.689      0.482
           diningtable        190        206      0.663      0.733      0.751      0.555
                   dog        418        489      0.848      0.476      0.705      0.517
                 horse        274        348      0.784      0.781      0.832      0.582
             motorbike        222        325      0.742      0.785      0.813      0.546
                person       2007       4528      0.819      0.774      0.844       0.56
           pottedplant        224        480      0.486      0.479      0.466      0.227
                 sheep         97        242      0.668      0.736      0.723      0.549
                  sofa        223        239      0.557      0.665      0.635      0.499
                 train        259        282      0.917      0.548       0.79      0.555
             tvmonitor        229        308      0.804      0.562      0.738      0.526
Speed: 0.1ms preprocess, 2.3ms inference, 0.0ms loss, 0.6ms postprocess per image
```

## COCO - Detect
I also tested COCO validation for pretrained models and they're the same.
**main**
```
In [4]: results = model.val(data="coco.yaml", save_json=True)
Ultralytics 8.3.48 🚀 Python-3.10.14 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15934MiB)
val: Scanning /datasets/coco/labels/val2017.cache... 4952 images, 48 backgrounds, 0 corrupt: 100%|██████████| 5000/5000 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 313/313 [00:43<00:00,  7.19it/s]
                   all       5000      36335      0.653      0.504      0.549      0.393
                person       2693      10777       0.76      0.676      0.753      0.524
               bicycle        149        314      0.692      0.404      0.478      0.276
                   car        535       1918      0.662      0.532      0.585      0.376
            motorcycle        159        367      0.765      0.599      0.674      0.444
              airplane         97        143      0.803      0.797      0.869      0.689
                   bus        189        283      0.772      0.707      0.773      0.651
                 train        157        190      0.827      0.811      0.863      0.681
                 truck        250        414      0.585      0.415      0.477      0.322
                  boat        121        424       0.58      0.351      0.413       0.22
         traffic light        191        634      0.647      0.345      0.416      0.218
          fire hydrant         86        101      0.867      0.743      0.806      0.639
             stop sign         69         75      0.712      0.613      0.662      0.601
         parking meter         37         60      0.698       0.55      0.613      0.467
                 bench        235        411      0.575       0.28      0.325      0.217
                  bird        125        427      0.618      0.382       0.44       0.28
                   cat        184        202      0.782      0.861      0.882      0.703
                   dog        177        218      0.761      0.743      0.779      0.648
                 horse        128        272      0.764      0.706       0.77      0.592
                 sheep         65        354        0.7      0.689      0.744      0.506
                   cow         87        372      0.719      0.679      0.732      0.522
              elephant         89        252      0.729      0.837      0.842      0.655
                  bear         49         71      0.851      0.859      0.911      0.745
                 zebra         85        266      0.804      0.838      0.893      0.673
               giraffe        101        232      0.859      0.845      0.891        0.7
              backpack        228        371      0.491      0.197      0.202      0.106
              umbrella        174        407      0.652      0.537      0.581      0.392
               handbag        292        540      0.476      0.128      0.165     0.0887
                   tie        145        252      0.635      0.417       0.45       0.28
              suitcase        105        299      0.636      0.472      0.536      0.375
               frisbee         84        115      0.787      0.783      0.814      0.613
                  skis        120        241      0.578      0.336      0.397      0.214
             snowboard         49         69      0.584      0.377      0.393      0.275
           sports ball        169        260      0.692      0.488      0.511      0.371
                  kite         91        327       0.64      0.541      0.599      0.412
          baseball bat         97        145      0.571      0.393      0.417      0.231
        baseball glove        100        148      0.682        0.5      0.539      0.323
            skateboard        127        179      0.706      0.648      0.678      0.467
             surfboard        149        267      0.645      0.498       0.54      0.344
         tennis racket        167        225      0.747      0.671      0.697      0.418
                bottle        379       1013      0.622      0.414      0.471      0.312
            wine glass        110        341      0.706      0.334      0.428       0.28
                   cup        390        895      0.581      0.448      0.501      0.363
                  fork        155        215      0.588      0.339      0.384      0.269
                 knife        181        325      0.462      0.151      0.202      0.122
                 spoon        153        253      0.398      0.154      0.162     0.0994
                  bowl        314        623      0.608      0.496      0.528      0.392
                banana        103        370      0.499      0.297      0.362      0.225
                 apple         76        236      0.399      0.237      0.235      0.167
              sandwich         98        177       0.56      0.503      0.483      0.363
                orange         85        285      0.473      0.399      0.389      0.301
              broccoli         71        312      0.498      0.356      0.385      0.213
                carrot         81        365      0.438      0.323      0.289      0.179
               hot dog         51        125      0.649      0.464      0.494      0.356
                 pizza        153        284      0.703      0.641       0.69      0.523
                 donut         62        328      0.611      0.507      0.549      0.427
                  cake        124        310      0.601      0.439      0.521      0.359
                 chair        580       1771      0.592      0.363       0.42      0.265
                 couch        195        261      0.623      0.582      0.632      0.476
          potted plant        172        342      0.549      0.395      0.395      0.228
                   bed        149        163      0.627      0.619      0.661      0.484
          dining table        501        695      0.545      0.449      0.449      0.312
                toilet        149        179      0.732      0.763      0.822      0.668
                    tv        207        288      0.721       0.67      0.725      0.548
                laptop        183        231      0.668       0.68      0.713      0.591
                 mouse         88        106      0.716      0.717      0.739      0.543
                remote        145        283      0.507      0.279      0.323       0.19
              keyboard        106        153      0.636      0.582      0.641      0.482
            cell phone        214        262      0.588      0.431       0.44      0.305
             microwave         54         55      0.681        0.6       0.67      0.521
                  oven        115        143      0.637       0.49      0.556      0.383
               toaster          8          9          1      0.427      0.691      0.455
                  sink        187        225      0.662      0.511      0.542      0.368
          refrigerator        101        126      0.769      0.643      0.724      0.574
                  book        230       1129        0.5      0.118      0.218      0.108
                 clock        204        267      0.755      0.625      0.657      0.462
                  vase        137        274      0.589      0.478      0.478      0.328
              scissors         28         36      0.529      0.306      0.327      0.276
            teddy bear         94        190      0.688      0.574      0.627      0.458
            hair drier          9         11          1          0    0.00103   0.000207
            toothbrush         34         57       0.47      0.281      0.252      0.165
Speed: 0.2ms preprocess, 2.6ms inference, 0.0ms loss, 1.1ms postprocess per image
Saving runs/detect/val10/predictions.json...

Evaluating pycocotools mAP using runs/detect/val10/predictions.json and /datasets/coco/annotations/instances_val2017.json...
loading annotations into memory...
Done (t=0.40s)
creating index...
index created!
Loading and preparing results...
DONE (t=5.10s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=51.34s).
Accumulating evaluation results...
DONE (t=9.89s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.394
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.553
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.428
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.199
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.432
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.570
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.324
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.540
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.598
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.370
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.663
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.781
Results saved to runs/detect/val10
```
**PR**
```
In [8]: model = YOLO("yolo11n.pt")

In [9]: results = model.val(data="coco.yaml", save_json=True)
Ultralytics 8.3.48 🚀 Python-3.10.14 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15934MiB)
YOLO11n summary (fused): 238 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Scanning /datasets/coco/labels/val2017.cache... 4952 images, 48 backgrounds, 0 corrupt: 100%|██████████| 5000/5000 [00:00<?, ?
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 313/313 [00:36<00:00,  8
                   all       5000      36335      0.653      0.504      0.549      0.393
                person       2693      10777       0.76      0.676      0.753      0.524
               bicycle        149        314      0.692      0.404      0.478      0.276
                   car        535       1918      0.662      0.532      0.585      0.376
            motorcycle        159        367      0.765      0.599      0.674      0.444
              airplane         97        143      0.803      0.797      0.869      0.689
                   bus        189        283      0.772      0.707      0.773      0.651
                 train        157        190      0.827      0.811      0.863      0.681
                 truck        250        414      0.585      0.415      0.477      0.322
                  boat        121        424       0.58      0.351      0.413       0.22
         traffic light        191        634      0.647      0.345      0.416      0.218
          fire hydrant         86        101      0.867      0.743      0.806      0.639
             stop sign         69         75      0.712      0.613      0.662      0.601
         parking meter         37         60      0.698       0.55      0.613      0.467
                 bench        235        411      0.575       0.28      0.325      0.217
                  bird        125        427      0.618      0.382       0.44       0.28
                   cat        184        202      0.782      0.861      0.882      0.703
                   dog        177        218      0.761      0.743      0.779      0.648
                 horse        128        272      0.764      0.706       0.77      0.592
                 sheep         65        354        0.7      0.689      0.744      0.506
                   cow         87        372      0.719      0.679      0.732      0.522
              elephant         89        252      0.729      0.837      0.842      0.655
                  bear         49         71      0.851      0.859      0.911      0.745
                 zebra         85        266      0.804      0.838      0.893      0.673
               giraffe        101        232      0.859      0.845      0.891        0.7
              backpack        228        371      0.491      0.197      0.202      0.106
              umbrella        174        407      0.652      0.537      0.581      0.392
               handbag        292        540      0.476      0.128      0.165     0.0887
                   tie        145        252      0.635      0.417       0.45       0.28
              suitcase        105        299      0.636      0.472      0.536      0.375
               frisbee         84        115      0.787      0.783      0.814      0.613
                  skis        120        241      0.578      0.336      0.397      0.214
             snowboard         49         69      0.584      0.377      0.393      0.275
           sports ball        169        260      0.692      0.488      0.511      0.371
                  kite         91        327       0.64      0.541      0.599      0.412
          baseball bat         97        145      0.571      0.393      0.417      0.231
        baseball glove        100        148      0.682        0.5      0.539      0.323
            skateboard        127        179      0.706      0.648      0.678      0.467
             surfboard        149        267      0.645      0.498       0.54      0.344
         tennis racket        167        225      0.747      0.671      0.697      0.418
                bottle        379       1013      0.622      0.414      0.471      0.312
            wine glass        110        341      0.706      0.334      0.428       0.28
                   cup        390        895      0.581      0.448      0.501      0.363
                  fork        155        215      0.588      0.339      0.384      0.269
                 knife        181        325      0.462      0.151      0.202      0.122
                 spoon        153        253      0.398      0.154      0.162     0.0994
                  bowl        314        623      0.608      0.496      0.528      0.392
                banana        103        370      0.499      0.297      0.362      0.225
                 apple         76        236      0.399      0.237      0.235      0.167
              sandwich         98        177       0.56      0.503      0.483      0.363
                orange         85        285      0.473      0.399      0.389      0.301
              broccoli         71        312      0.498      0.356      0.385      0.213
                carrot         81        365      0.438      0.323      0.289      0.179
               hot dog         51        125      0.649      0.464      0.494      0.356
                 pizza        153        284      0.703      0.641       0.69      0.523
                 donut         62        328      0.611      0.507      0.549      0.427
                  cake        124        310      0.601      0.439      0.521      0.359
                 chair        580       1771      0.592      0.363       0.42      0.265
                 couch        195        261      0.623      0.582      0.632      0.476
          potted plant        172        342      0.549      0.395      0.395      0.228
                   bed        149        163      0.627      0.619      0.661      0.484
          dining table        501        695      0.545      0.449      0.449      0.312
                toilet        149        179      0.732      0.763      0.822      0.668
                    tv        207        288      0.721       0.67      0.725      0.548
                laptop        183        231      0.668       0.68      0.713      0.591
                 mouse         88        106      0.716      0.717      0.739      0.543
                remote        145        283      0.507      0.279      0.323       0.19
              keyboard        106        153      0.636      0.582      0.641      0.482
            cell phone        214        262      0.588      0.431       0.44      0.305
             microwave         54         55      0.681        0.6       0.67      0.521
                  oven        115        143      0.637       0.49      0.556      0.383
               toaster          8          9          1      0.427      0.691      0.455
                  sink        187        225      0.662      0.511      0.542      0.368
          refrigerator        101        126      0.769      0.643      0.724      0.574
                  book        230       1129        0.5      0.118      0.218      0.108
                 clock        204        267      0.755      0.625      0.657      0.462
                  vase        137        274      0.589      0.478      0.478      0.328
              scissors         28         36      0.529      0.306      0.327      0.276
            teddy bear         94        190      0.688      0.574      0.627      0.458
            hair drier          9         11          1          0    0.00103   0.000207
            toothbrush         34         57       0.47      0.281      0.252      0.165
Speed: 0.1ms preprocess, 2.3ms inference, 0.0ms loss, 0.7ms postprocess per image
Saving runs/detect/val8/predictions.json...

Evaluating pycocotools mAP using runs/detect/val8/predictions.json and /datasets/coco/annotations/instances_val2017.json...
loading annotations into memory...
Done (t=0.40s)
creating index...
index created!
Loading and preparing results...
DONE (t=4.91s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=50.86s).
Accumulating evaluation results...
DONE (t=9.70s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.394
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.553
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.428
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.199
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.432
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.570
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.324
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.540
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.598
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.370
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.663
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.781
Results saved to runs/detect/val8
```

## COCO - Pose
**main**
```
In [2]: model = YOLO("yolo11n-pose.pt")

In [3]: results = model.val(data="coco-pose.yaml", save_json=True)
Ultralytics 8.3.48 🚀 Python-3.10.14 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15934MiB)
YOLO11n-pose summary (fused): 257 layers, 2,866,468 parameters, 0 gradients, 7.4 GFLOPs
val: Scanning /datasets/coco-pose/labels/val2017.cache... 2346 images, 0 backgrounds, 0 corrupt: 100%|██████████| 2346/2346 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Pose(P          R      mAP50  mAP50-95): 100%|██████████| 147/147 [00:18<0
                   all       2346       6352       0.89      0.826      0.922       0.71      0.849       0.76       0.81      0.511
Speed: 0.1ms preprocess, 2.6ms inference, 0.0ms loss, 0.6ms postprocess per image
Saving runs/pose/val7/predictions.json...

Evaluating pycocotools mAP using runs/pose/val7/predictions.json and /datasets/coco-pose/annotations/person_keypoints_val2017.json...
loading annotations into memory...
Done (t=0.20s)
creating index...
index created!
Loading and preparing results...
DONE (t=1.95s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=7.10s).
Accumulating evaluation results...
DONE (t=0.54s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.533
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.719
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.588
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.205
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.658
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.799
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.210
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.549
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.621
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.233
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.749
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.865
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=7.60s).
Accumulating evaluation results...
DONE (t=0.14s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.500
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.806
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.535
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.436
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.609
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.577
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.859
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.622
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.496
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.689
Results saved to runs/pose/val7
```
**PR**
```
In [12]: model = YOLO("yolo11n-pose.pt")

In [13]: results = model.val(data="coco-pose.yaml", save_json=True)
Ultralytics 8.3.48 🚀 Python-3.10.14 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15934MiB)
YOLO11n-pose summary (fused): 257 layers, 2,866,468 parameters, 0 gradients, 7.4 GFLOPs
val: Scanning /datasets/coco-pose/labels/val2017.cache... 2346 images, 0 backgrounds, 0 corrupt: 100%|██████████| 2346/2346 [00:00<
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Pose(P          R      mAP50  mAP50-95
                   all       2346       6352       0.89      0.826      0.922       0.71      0.849       0.76       0.81      0.511
Speed: 0.1ms preprocess, 2.6ms inference, 0.0ms loss, 0.6ms postprocess per image
Saving runs/pose/val6/predictions.json...

Evaluating pycocotools mAP using runs/pose/val6/predictions.json and /datasets/coco-pose/annotations/person_keypoints_val2017.json...
loading annotations into memory...
Done (t=0.20s)
creating index...
index created!
Loading and preparing results...
DONE (t=2.19s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=7.05s).
Accumulating evaluation results...
DONE (t=0.56s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.533
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.719
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.588
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.205
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.658
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.799
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.210
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.549
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.621
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.233
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.749
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.865
Running per image evaluation...
Evaluate annotation type *keypoints*
DONE (t=7.62s).
Accumulating evaluation results...
DONE (t=0.16s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.500
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.806
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.535
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.436
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.609
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.577
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.859
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.622
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.496
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.689
Results saved to runs/pose/val6
```

**DOTAv1 - OBB**

There's an error when running `model.val()` with `save_json=True` on `DOTAv1.yaml`. But this error also exists on the `main` branch:
```
In [4]: model = YOLO("yolo11n-obb.pt")

In [5]: results = model.val(data="DOTAv1.yaml", save_json=True)
Ultralytics 8.3.48 🚀 Python-3.10.14 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15934MiB)
YOLO11n-obb summary (fused): 257 layers, 2,656,648 parameters, 0 gradients, 6.6 GFLOPs
val: Scanning /datasets/DOTAv1/labels/val.cache... 458 images, 2 backgrounds, 0 corrupt: 100%|██████████| 458/458 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 29/29 [00:18<00:00,  1.60it/s]
                   all        458      28853      0.794      0.561      0.618      0.472
                 plane         70       2531      0.923      0.734      0.809       0.66
                  ship        108       8960      0.879      0.608        0.7       0.53
          storage tank         55       2888      0.913      0.378      0.525      0.378
      baseball diamond         53        214      0.834      0.682      0.742      0.554
          tennis court         94        760      0.955      0.905       0.94      0.902
      basketball court         35        132      0.703      0.439      0.507       0.45
    ground track field         70        144       0.67      0.528      0.558      0.448
                harbor        114       2090       0.83      0.814      0.844      0.521
                bridge         75        464      0.677      0.258      0.318      0.135
         large vehicle        110       4387      0.881       0.82      0.861      0.689
         small vehicle        136       5438      0.701       0.59      0.648      0.478
            helicopter         14         73      0.833      0.478      0.513       0.39
            roundabout         61        179      0.728      0.279      0.335      0.239
     soccer ball field         63        153      0.653      0.438      0.458      0.393
         swimming pool         41        440      0.734      0.464      0.512      0.319
Speed: 0.5ms preprocess, 6.8ms inference, 0.0ms loss, 4.7ms postprocess per image
Saving runs/obb/val5/predictions.json...
Saving predictions with DOTA format to runs/obb/val5/predictions_txt...
Saving merged predictions with DOTA format to runs/obb/val5/predictions_merged_txt...
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[5], line 1
----> 1 results = model.val(data="DOTAv1.yaml", save_json=True)

File /ultralytics/ultralytics/engine/model.py:639, in Model.val(self, validator, **kwargs)
    636 args = {**self.overrides, **custom, **kwargs, "mode": "val"}  # highest priority args on the right
    638 validator = (validator or self._smart_load("validator"))(args=args, _callbacks=self.callbacks)
--> 639 validator(model=self.model)
    640 self.metrics = validator.metrics
    641 return validator.metrics

File /opt/conda/lib/python3.10/site-packages/torch/utils/_contextlib.py:116, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    113 @functools.wraps(func)
    114 def decorate_context(*args, **kwargs):
    115     with ctx_factory():
--> 116         return func(*args, **kwargs)

File /ultralytics/ultralytics/engine/validator.py:217, in BaseValidator.__call__(self, trainer, model)
    215         LOGGER.info(f"Saving {f.name}...")
    216         json.dump(self.jdict, f)  # flatten and save
--> 217     stats = self.eval_json(stats)  # update stats
    218 if self.args.plots or self.args.save_json:
    219     LOGGER.info(f"Results saved to {colorstr('bold', self.save_dir)}")

File /ultralytics/ultralytics/models/yolo/obb/val.py:177, in OBBValidator.eval_json(self, stats)
    175 image_id = d["image_id"].split("__")[0]
    176 pattern = re.compile(r"\d+___\d+")
--> 177 x, y = (int(c) for c in re.findall(pattern, d["image_id"])[0].split("___"))
    178 bbox, score, cls = d["rbox"], d["score"], d["category_id"]
    179 bbox[0] += x

IndexError: list index out of range
```

This probably needs to fixed separately.

Laughing-q commented:
@Y-T-G Nice work here!
@glenn-jocher @Y-T-G Guys I feel like we should be more careful of changing the default behavior of our package. I remember we have this index started from 0 ever since @glenn-jocher created yolov5 repo, some users might already get used to this and have some extra "postprocess" embedded in their workflow to handle it if they want it start from index 1, and now we are updating this default behavior silently which might break their workflow.

Y-T-G commented:
@Laughing-q True. Maybe we can have this behavior when someone passes `save_json="coco"`, while `save_json=True` has the original behavior.

Laughing-q commented:
@Y-T-G hmmm probably just keep it as it is now. Just be careful for future changes when it comes to default behavior.
